### ISSUE TITLE
Static size boolean masking

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#96111]
### 🐛 Describe the bug

A long standing request is https://github.com/pytorch/pytorch/issues/62320 ; Executorch team has agreed to implement it.

Once this is implemented, we can also get static size boolean masking to work too. The easiest way is to convert the boolean mask into an index tensor. You can use the meta implementation for indexing to do this:

```
@register_meta(aten.index.Tensor)
def meta_index_Tensor(self, indices):
    result: List[Optional[Tensor]] = []
    for i, index in enumerate(indices):
        if index is not None:
            check(
                index.dtype in [torch.long, torch.int, torch.int8, torch.bool],
                lambda: "tensors used as indices must be long, int, byte or bool tensors",
            )
            if index.dtype in [torch.int8, torch.bool]:
                nonzero = index.nonzero()
                k = len(result)
                check(
                    k + index.ndim <= self.ndim,
                    lambda: f"too many indices for tensor of dimension {self.ndim}",
                    IndexError,
                )
                for j in range(index.ndim):
                    check(
                        index.shape[j] == self.shape[k + j],
                        lambda: f"The shape of the mask {index.shape} at index {i} "
                        f"does not match the shape of the indexed tensor {self.shape} at index {k + j}",
                        IndexError,
                    )
                    result.append(nonzero.select(1, j))
            else:
                result.append(index)
        else:
            result.append(index)
    return result

```

The main annoyance is that if there are not enough elements to fill the nonzero, it will be zero padded. For a boolean mask, a zero pad is inappropriate; instead you want an invalid index, and then to fill the indexing op with some placeholder element like 0.

### Versions

master

**Comment by vadimkantorov** [comment]
btw, about nonzero() - also now it can be done in batched way returning nestedtensor / paddedsequence (or tensorlist - which i hope will also be introduced one day)

**Comment by suo** [comment]
who agreed to do this? @tugsbayasgalan do you know

**Comment by guangy10** [comment]
@ezyang At high-level, is the static boolean masking an equivalent to [tf.boolean_mask](https://www.tensorflow.org/api_docs/python/tf/boolean_mask)? 
Curious if it's mainly needed by export or there are other use cases?
Would you mind further elaborating why convert the boolean mask into an index tensor will be helpful?

**Comment by ezyang** [comment]
Yes but static.

We were predicting that static size boolean masking may be necessary because most models don't use nonzero directly, they indirectly use it via boolean masking.

Converting it to an index tensor is helpful because this is a decomposition, that works so long as you have nonzero_static.

**Comment by vadimkantorov** [comment]
@ezyang if you mean an op that produces somehow a mask output for nonzero, one option is to make it return a sparse tensor with ones at the nonzero places, it is somehow related to https://github.com/pytorch/pytorch/issues/34951 and a bit to https://github.com/pytorch/pytorch/issues/62162
